MetaMask/metamask-mobile
# feat: Use new Segment implementation in onboarding steps (Batch 3 of 4)

## **Description**

1. What is the reason for the change?
  Implement the New Segment metrics system created in PR #7520 in the onboarding steps.

2. What is the improvement/solution?
  - Initialise MetaMetrics on app start without removing Analytics yet as we still uses it outside of onboarding
  - create a `trackAfterInteractions` utility function to reduce code when tracking events.
  - Replace calls to Analytics with calls to `trackAfterInteractions` in all components involved in onboarding screens (see [`OnboardingNav`](https://github.com/MetaMask/metamask-mobile/blob/feat%2F1275_segment_onboarding_impl/app/components/Nav/App/index.js#L122-L175) in `app/components/Nav/App/index.js`)
  - Fix onboarding events queue usage in `app/components/UI/OptinMetrics/index.js`
  - Update Opt-on/Opt-out in onboarding
  - Extract Analytics User profil properties in UserProfileMetaData util and inject it in MetaMetrics identify request alongside device metadata

## **Related issues**

Fixes https://github.com/MetaMask/mobile-planning/issues/1275

## **Manual testing steps**

```gherkin
Scenario Outline: Onboarding up to metrics opt-in

Given user installed Metamask on the device

And user starts the Metamask for the first time

And user sees the the “Welcome to MetaMask” screen

And user touches “Get Started” button

And user touches <wallet action>

And user sees the “Help us improve Metamask” screen

When user touches “I agree” button

Then events (from app start) appear on Segment Dashboard source(1) debugger after at most 30s

And events are the following (not necessarily in order):

- `identify` type event with userId of the form of a UUIDv4
- `Welcome Message Viewed`
- `Onboarding Started`
- `Welcome Screen Engagement` (more than one if swiped through the welcome screen images)
- `Wallet Setup Started`
- `Analytics Preference Selected` with `properties[analytics_option_selected]=Metrics Opt In`

Examples:
| wallet action | 
|    “create new wallet” | 
|    “Import Using Recovery Phrase” | 
```

```gherkin
Scenario: Finish Onboarding after Opt-in with new wallet

Given user started a new wallet onboarding

And user touched “I agree” button on “Help us improve Metamask” screen

And events tracked up to the opt-in are displayed on Segment Dashboard source debugger

And user continues the onboarding

When user reaches the wallet screen at the end of onboarding

Then Segment Dashboard source debugger contains the following events:

- `Wallet Creation Attempted`
- `Wallet Created`
- `Wallet Setup Completed` with properties `{"new_wallet": true, "wallet_setup_type": "new"}`

And if securing SRP is skipped:

- `Wallet Security Skip Initiated`
- `Wallet Security Skip Confirmed`

But if securing SRP is done:

- `Wallet Security Started`
- `Manual Backup Initiated`
- `Phrase Revealed`
- `Phrase Confirmed`
- `Wallet Security Completed`
```

```gherkin
Scenario: Finish Onboarding after Opt-in with imported wallet

Given user started an imported wallet onboarding

And user touched “I agree” button on “Help us improve Metamask” screen

And events tracked up to the opt-in are displayed on Segment Dashboard source debugger

And user continues the onboarding

When user reaches the wallet screen at the end of onboarding

Then Segment Dashboard source debugger contains the following events:

- `Wallet Import Started`
- `Wallet Import Attempted`

And if user fails some import attempt you should have some occurrences of:

- `Wallet Setup Failure`

And when import SRP succeeds:

- `Wallet Imported`
- `Wallet Setup Completed`
- `Wallet Security Completed`
```

```gherkin
Scenario Outline: Onboarding with metrics opt-out

Given user installed Metamask on the device

And user starts the Metamask for the first time

And user sees the the “Welcome to MetaMask” screen

And user touches “Get Started” button

And user touches <wallet action>

And user sees the “Help us improve Metamask” screen

And user touches “No thanks” button

When user finishes the onboarding up to the wallet screen

Then no event at all appear on Segment Dashboard source(1) debugger even after more than 30s

Examples:
| wallet action | 
|    “create new wallet” | 
|    “Import Using Recovery Phrase” |
```

(1) source to look at should match the app type: [dev](https://app.segment.com/consensys-analytics/sources/raw_segment_metamask_mobile_dev/debugger)/[qa](https://app.segment.com/consensys-analytics/sources/raw_segment_metamask_mobile_qa/debugger). Ask in Slack if you need access to Segment dashboards.

## **Screenshots/Recordings**

The event tracking scope for this PR is all the screens between first start welcome screen:

<img width="100" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/8524aa5b-3e82-499c-bcb2-852f6adf5dae">

and wallet screen:

<img width="100" alt="image" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/54172c4b-53da-4e7a-9994-d3cf3748c796">


From 1st start to optin. The dev logs should display the events tracked.
Notice the delay before they are sent as we have a 30s batch timer that actually sends the events in batches either every 30s or when the batch reaches 20 events.

https://github.com/MetaMask/metamask-mobile/assets/4677568/8c71bfe8-efc8-40c7-9ebc-961c2cb6aa00

Then you can see the events on the Segment Dashboard. Best is to filter either by date or if you have the logs, copy the user id.

<img width="1476" alt="Screenshot 2023-12-06 at 18 30 27" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/d3dda082-8b70-4a5e-bd01-30a10be793cb">

Onboarding started event:
<img width="1124" alt="Screenshot 2023-12-06 at 18 31 40" src="https://github.com/MetaMask/metamask-mobile/assets/4677568/e7731fc1-ada6-4f01-bd4c-4d4a5ce2f3d5">

> [!NOTE]  
> Events may not display in chronological order currently due to a behaviour on Segment server side that rewrites the timestamps. This is being worked on bit not in this PR.

### **Before**

No visual change for users

### **After**

No visual change for users

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
